contrib/mypy/src/python/pants/contrib/mypy/tasks/mypy_task.py

```diff
@@ -0,0 +1,132 @@
+# coding=utf-8
+# Copyright 2017 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+from __future__ import (absolute_import, division, generators, nested_scopes, print_function,
+                        unicode_literals, with_statement)
+
+import os
+
+from pants.backend.python.interpreter_cache import PythonInterpreter, PythonInterpreterCache
+from pants.backend.python.python_requirement import PythonRequirement
+from pants.backend.python.targets.python_binary import PythonBinary
+from pants.backend.python.targets.python_library import PythonLibrary
+from pants.backend.python.targets.python_target import PythonTarget
+from pants.backend.python.targets.python_tests import PythonTests
+from pants.backend.python.tasks.python_task import PythonTask
+from pants.base.build_environment import get_buildroot
+from pants.base.exceptions import TaskError
+from pants.base.workunit import WorkUnit, WorkUnitLabel
+from pants.util.contextutil import temporary_dir
+from pants.util.process_handler import subprocess
+from pex.pex import PEX
+from pex.pex_builder import PEXBuilder
+from pex.pex_info import PexInfo
+
+
+class MypyTask(PythonTask):
+  """Invoke the mypy static type analyzer for Python."""
+
+  _PYTHON_SOURCE_EXTENSION = '.py'
+
+  @classmethod
+  def register_options(cls, register):
+    register('--mypy-version', default='0.550', help='The version of mypy to use')
+
+  @classmethod
+  def supports_passthru_args(cls):
+    return True
+
+  def py3_interpreter(self):
+    interpreters = self._interpreter_cache.setup(filters=['CPython>=3'])
+    return min(interpreters) if interpreters else None
+
+  @staticmethod
+  def is_non_synthetic_python_target(target):
+    return not target.is_synthetic and isinstance(target, (PythonLibrary, PythonBinary, PythonTests))
+
+  @staticmethod
+  def is_python_target(target):
+    return isinstance(target, (PythonTarget,))
+
+  def _calculate_python_sources(self, targets):
+    """Generate a set of source files from the given targets."""
+    python_eval_targets = filter(self.is_non_synthetic_python_target, targets)
+    sources = set()
+    for target in python_eval_targets:
+      sources.update(
+        source for source in target.sources_relative_to_buildroot()
+        if os.path.splitext(source)[1] == self._PYTHON_SOURCE_EXTENSION
+      )
+    return list(sources)
+
+  def _calculate_python_package_roots(self, targets):
```

Comment:
Doesn't `{target.target_base for target in targets}` give you what you want?

Reply:
Yeah I switched to that after getting that from John on Slack general.